File: MergeTwoSortedLists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode()
        ans = dummy


        while list1 or list2:
            if list1 and list2:
                if list1.val < list2.val:
                    dummy.next = ListNode(list1.val, None)
                    dummy = dummy.next
                    list1 = list1.next
                else:
                    dummy.next = ListNode(list2.val, None)
                    dummy = dummy.next
                    list2 = list2.next 
            elif not list1:
                dummy.next = list2
                list2 = None
            else:
                dummy.next = list1
                list1 = None
        
        return ans.next

    #Time: O(n + m)
    #Space: O(1)

File: MergeTwoSortedLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        ListNode dummy = new ListNode();
        ListNode ans = dummy;

        while(list1 != null || list2 != null){
            if (list1 != null && list2 != null ){
                if (list1.val < list2.val){
                    dummy.next = new ListNode(list1.val, null);
                    list1 = list1.next;
                }else{
                    dummy.next = new ListNode(list2.val, null);
                    list2 = list2.next;
                }
                dummy = dummy.next;
            }
            else if(list1 == null){
                dummy.next = list2;
                list2 = null;
            }else{
                dummy.next = list1;
                list1 = null;
            }
        }
        return ans.next;
    }
    //Time: O(n + m)
    //Space: O(1)
}
